### PR TITLE
Prefer using broadcast_in_dim/squeeze instead of reshape

### PR DESCRIPTION
**Pull request by shoyer**
`reshape()` is quite powerful, but does not necessarily preserve a notion of axis identity (particularly for axes of length 1). This is problematic for transformation rules that need to preserve a notion of axis identity, such as for masking and a new transformation rule I'm exploring for unraveling pytrees.

This PR adds a new `lax.squeeze` primitive, and rewrite functions from `lax_numpy.py` in terms of `lax.squeeze`/`lax.broadcast_in_dim`, whenever possible, to ensure a well-defined mapping between input and output axes. In particular, it updates: `matmul`, various `stack` functions, the `array` constructor, broadcasting arithmetic, array indexing, `squeeze` and reductions with `keepdims=True`.

I also implemented support for multiple axes in `jnp.expand_dims` (added in NumPy 1.18), since it was convenient for some of these other functions.

I considered trying to write a masking rule for broadcast_in_dim as well, but it was trickier than I expected and @JuliusKunze has probably already thought about it :)

### CODE REVIEW

**Comment by JuliusKunze** [comment]
@shoyer Indeed! I have the masking rule for `broadcast_in_dim` implemented locally, waiting for https://github.com/google/jax/pull/2922 to be merged before I can PR.

File: tests/lax_numpy_test.py [hunk 1]
```diff
@@ -1811,8 +1811,11 @@ def testReshapeMethod(self, arg_shape, out_shape, dtype, rng_factory):
        "rng_factory": jtu.rand_default}
       for arg_shape in [(), (3,), (3, 4)]
       for dtype in default_dtypes
-      for dim in range(-len(arg_shape)+1, len(arg_shape))))
+      for dim in (list(range(-len(arg_shape)+1, len(arg_shape)))
+                  + [(0,), (len(arg_shape), len(arg_shape) + 1)])))
   def testExpandDimsStaticDim(self, arg_shape, dtype, dim, rng_factory):
+    if isinstance(dim, tuple) and numpy_version < (1, 18, 0):
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py:1817]: I think you only need to skip the `CheckAgainstNumpy` test, but you can leave the `CompileAndCheck` part of the test enabled.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -1153,29 +1152,38 @@ def unravel_index(indices, shape):
...
+    if not axis_set:
+      return a
+    if _any(shape_a[a] != 1 for a in axis_set):
       raise ValueError("cannot select an axis to squeeze out which has size "
                        "not equal to one")
-    newshape = [d for i, d in enumerate(shape_a)
-                if d != 1 or i not in axis]
-  return lax.reshape(a, newshape)
+    indices = tuple(0 if i in axis_set else slice(None) for i in range(len(shape_a)))
+  # TODO: consider using lax.slice instead here?
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:1171]: It seems to me like it is using too heavyweight a primitive (`gather`) to collapse away degenerate dimensions. It's pretty complicated to understand gather's semantics and I'm not even completely sure XLA will optimize away this kind of degenerate gather. (Hopefully it will.)

Perhaps we should just add a new lax-level `squeeze` primitive? You could implement this via `Reshape` in its translation rule, but it would preserve the necessary structure at the `lax` level.

File: jax/numpy/lax_numpy.py [hunk 3]
```diff
@@ -2151,7 +2163,7 @@ def array(object, dtype=None, copy=True, order="K", ndmin=0):
     raise TypeError("Unexpected input type for array: {}".format(type(object)))
 
   if ndmin > ndim(out):
-    out = lax.reshape(out, (1,) * (ndmin - ndim(out)) + shape(out))
+    out = broadcast_to(out, (1,) * (ndmin - ndim(out)) + shape(out))
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:2166]: For broadcasting in leading dimensions you could simply write:
`lax.broadcast(out, (1,) * (ndmin - ndim(out))`

File: jax/numpy/lax_numpy.py [hunk 4]
```diff
@@ -2394,7 +2406,7 @@ def ix_(*args):
       # Numpy uses an integer index type for empty arrays.
       output.append(lax.full(shape, np.zeros((), np.intp)))
     else:
-      output.append(lax.reshape(a, shape))
+      output.append(lax.broadcast_in_dim(a, shape, [i]))
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:2409]: I'd probably write `(i,)` given `lax` ends up converting to a tuple.

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (shoyer) [reply jax/numpy/lax_numpy.py]: Good idea! I added `lax.squeeze`, with an implementation in terms of `reshape`. There's a nice duality between `squeeze` and `broadcast_in_dim` (for insertion of size 1 dimensions).

File: jax/lax/lax.py [hunk 5]
```diff
@@ -2922,6 +2922,71 @@ def _pad_translation_rule(c, operand, padding_value, *, padding_config):
 batching.primitive_batchers[pad_p] = _pad_batch_rule
 
 
+# the squeeze primitive exists for the benefit of batching and other rules that
```
Comment (hawkinsp) [inline jax/lax/lax.py:2925]: You might mention why it is preferable as opposed to `reshape`, perhaps with a small example.

File: jax/lax/lax.py [hunk 6]
```diff
@@ -2922,6 +2922,71 @@ def _pad_translation_rule(c, operand, padding_value, *, padding_config):
...
+  return tuple(s for i, s in enumerate(shape) if i not in dims_set)
+
+def _squeeze_translation_rule(c, arg, *, dimensions):
+  new_shape = _compute_squeeze_shape(c.get_shape(arg).dimensions(), dimensions)
+  return xops.Reshape(arg, new_shape)
+
+def _squeeze_transpose_rule(t, operand, *, dimensions):
+  assert ad.is_undefined_primal(operand)
+  shape = list(t.shape)
+  for d in sorted(dimensions):
```
Comment (hawkinsp) [inline jax/lax/lax.py:2961]: Could we just call `expand_dims` here?

File: jax/numpy/lax_numpy.py [hunk 7]
```diff
@@ -1153,29 +1152,21 @@ def unravel_index(indices, shape):
...
-    if _any(shape_a[a] != 1 for a in axis):
-      raise ValueError("cannot select an axis to squeeze out which has size "
-                       "not equal to one")
-    newshape = [d for i, d in enumerate(shape_a)
-                if d != 1 or i not in axis]
-  return lax.reshape(a, newshape)
+    axis = tuple(i for i, d in enumerate(shape_a) if d == 1)
+  elif isinstance(axis, int):
+    axis = (axis,)
+  axis = tuple(_canonicalize_axis(i, ndim(a)) for i in axis)
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py:1161]: Redundant with the canonicalization in `lax.squeeze`?

**Review by hawkinsp — APPROVED** [review]
Nice cleanup!

**Review by shoyer — COMMENTED** [review]
@hawkinsp please take another look when you get the chance

File: jax/lax/lax.py [hunk 8]
```diff
@@ -681,6 +682,10 @@ def reshape(operand: Array, new_sizes: Shape,
   """Wraps XLA's `Reshape
   <https://www.tensorflow.org/xla/operation_semantics#reshape>`_
   operator.
+
+  For inserting/removing dimensions of size 1, prefer using ``lax.squeeze`` /
+  ``lax.expand_dims``. These preserve information about axis identity that may
+  be useful for advanced transformation rules.
```
Comment (shoyer) [inline jax/lax/lax.py:688]: I'm a little conflicted about whether it's worth adding this to the docstring or not. Might be better just to keep it as a code comment.

**Review by hawkinsp — APPROVED** [review]

**Comment by shoyer** [comment]
Thanks for the quick review!
